File: 054_CountGoodNodesInBinaryTree.cpp
// Topic: Count Good Nodes in Binary Tree
// Description: Count nodes not smaller than any ancestor on path from root.
// Approach: DFS passing current maximum along path; increment when node->val >= current max.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildLevel(const vector<string>& vals)
{
    if (vals.empty() || vals[0] == "null")
        return nullptr;
    TreeNode* root = new TreeNode(stoi(vals[0]));
    queue<TreeNode*> q;
    q.push(root);
    int i = 1;
    while (!q.empty() && i < (int)vals.size())
    {
        TreeNode* node = q.front();
        q.pop();
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->left = new TreeNode(stoi(vals[i]));
            q.push(node->left);
        }
        ++i;
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->right = new TreeNode(stoi(vals[i]));
            q.push(node->right);
        }
        ++i;
    }
    return root;
}

int goodNodes(TreeNode* root)
{
    int ans = 0;
    function<void(TreeNode*, int)> dfs = [&](TreeNode* node, int maxSoFar)
    {
        if (!node)
            return;
        if (node->val >= maxSoFar)
        {
            ++ans;
            maxSoFar = node->val;
        }
        dfs(node->left, maxSoFar);
        dfs(node->right, maxSoFar);
    };
    dfs(root, INT_MIN);
    return ans;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<string> vals1 = {"3", "1", "4", "3", "null", "1", "5"};
    vector<string> vals2 = {"3", "3", "null", "4", "2"};
    TreeNode* r1 = buildLevel(vals1);
    TreeNode* r2 = buildLevel(vals2);
    cout << "Test 1 -> " << goodNodes(r1) << "\n";
    cout << "Test 2 -> " << goodNodes(r2) << "\n";
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/

File: 054_CountGoodNodesInBinaryTree.py
# Topic: Count Good Nodes in Binary Tree
# Description: Count nodes not smaller than any ancestor on path from root.
# Approach: DFS passing current maximum along path; increment when node.val >= current max.

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, max_so_far):
            if not node:
                return 0
            count = 1 if node.val >= max_so_far else 0
            max_so_far = max(max_so_far, node.val)
            count += dfs(node.left, max_so_far)
            count += dfs(node.right, max_so_far)
            return count

        return dfs(root, float('-inf'))


if __name__ == "__main__":
    # Test 1
    root = TreeNode(3, TreeNode(1, TreeNode(3)), TreeNode(4, TreeNode(1), TreeNode(5)))
    print(f"Test 1 -> {Solution().goodNodes(root)}")

    # Test 2
    root = TreeNode(3, TreeNode(3, TreeNode(4), TreeNode(2)))
    print(f"Test 2 -> {Solution().goodNodes(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""
